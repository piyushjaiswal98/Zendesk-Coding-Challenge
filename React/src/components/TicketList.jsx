import React, { useEffect, useState } from 'react';
import './TicketList.scss';

const TicketList = (props) => {

    const [data,setData] = useState([]);

    useEffect(() => {
        setData(props.ticketList);
    },[props.ticketList]);

    function handleExpand(id){
        var height = document.getElementById('ticket'+id).offsetHeight;
        if(height === 80)
        {
            document.getElementById('ticket'+id).style.height = "180px";
            document.getElementById('description'+id).style.visibility = "visible";
        }
        else
        {
            document.getElementById('ticket'+id).style.height = "80px";
            document.getElementById('description'+id).style.visibility = "hidden";
        }
    }

    function handleUrl(event){
        event.stopPropagation();
    }

    return(

        <div className="ticketList">
            {
                data.length>0 &&
                    data.map((d) => (
                        <div key={d.id} className={`ticket ticket${d.id}`} id={`ticket${d.id}`} onClick={e => handleExpand(d.id)}>
                            <div className="divbox">
                                <div className="divbox-1">
                                    <p>#{d.id}&nbsp;&nbsp;{d.subject}</p>
                                </div>
                                <div className="divbox-2">
                                    <div className={`status ${d.status}`}>{d.status}</div>
                                </div>
                            </div>
                            <div className="url" onClick={handleUrl}>
                                <p id="url">URL : <a target="_blank" rel="noreferrer" href={d.url}>Get Ticket JSON</a></p>
                            </div>

                            
                            <div className="description" id={`description${d.id}`}>
                                {d.description}
                            </div>
                            

                        </div>
                    ))
            }
        </div>
    )

}

export default TicketList;