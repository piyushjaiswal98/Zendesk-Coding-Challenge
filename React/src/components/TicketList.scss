.ticketList{
    height: inherit;
    width: inherit;
    padding: 1%;
    overflow: auto;
    .ticket{
        width: 99%;
        height: 80px;
        margin-bottom: 10px;
        background-color: rgb(89, 101, 119);
        border-radius: 5px;
        cursor: pointer;
        .divbox{
            height: 40px;
            display: flex;
            flex-direction: row;
            .divbox-1{
                flex: 0 0 80%;
                p{
                    margin: 5px 0 0 15px;
                    text-align: left;
                }
            }
            .divbox-2{
                flex: 0 0 20%;
                // padding: 10%;
                .status{
                    // padding: 5px;
                    margin-top: 5px;
                    margin-left: 15%;
                    width: 70%;
                    // padding-bottom: 45px;
                    height: 100%;
                    border-radius: 10px;
                }
                .open{
                    background-color: rgb(236, 70, 70);
                }
                .pending{
                    background-color: #61dafb;
                }
                .solved{
                    background-color: #aaaaaa;
                }
            }
        }
        
        .url{
            width: fit-content;
            #url{
                margin: 3px 0 0 56px;
                padding:0;
                text-align: left;
                font-size: 0.8em;
                z-index: 99;
                a{
                    text-decoration: none;
                    color: #61dafb;
                }
            }
        }
        
        .description{
            width: 90%;
            margin-left: 5%;
            margin-top: 7px;
            height: 90px;
            overflow: auto;
            font-size: 0.8em;
            text-align: left;
            padding: 1px 10px 1px 10px;
            visibility: hidden;
            border: 1px solid #61dafb;
            border-radius: 5px;
            background-color: white;
            color: black;
        }
    }
    .expand{
        height: 180px;
    }
}