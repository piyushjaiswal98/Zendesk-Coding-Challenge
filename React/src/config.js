module.exports = global.config = {
    uri: process.env.REACT_APP_API_HOST || 'http://localhost:2000/'
};