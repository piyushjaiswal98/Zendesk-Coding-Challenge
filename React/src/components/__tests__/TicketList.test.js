import { render, screen, cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';
import TicketList from '../TicketList';
import '../../config'

afterEach(()=> {
    cleanup();
})

const data = [{ 
    "ticket": { 
        "url": "https://zccstudents3001.zendesk.com/api/v2/tickets/4.json", 
        "id": 4, 
        "created_at": "2021-11-24T21:20:18Z", 
        "updated_at": "2021-11-24T21:20:18Z", 
        "subject": "ad sunt qui aute ullamco", 
        "raw_subject": "ad sunt qui aute ullamco", 
        "description": "Sunt incididunt officia proident elit anim ullamco reprehenderit ut. Aliqua sint amet aliquip cillum minim magna consequat excepteur fugiat exercitation est exercitation. Adipisicing occaecat nisi aliqua exercitation.\n\nAute Lorem aute tempor sunt mollit dolor in consequat non cillum irure reprehenderit. Nulla deserunt qui aliquip officia duis incididunt et est velit nulla irure in fugiat in. Deserunt proident est in dolore culpa mollit exercitation ea anim consequat incididunt. Mollit et occaecat ullamco ut id incididunt laboris occaecat qui.", 
        "status": "open", 
    } 
}];

test('Should render TicketList Component', () => {
    render(<TicketList ticketList={data}/>);
    const TicketListElement = screen.getByTestId('TicketList-1');
    expect(TicketListElement).toBeInTheDocument();
    expect(TicketListElement).toHaveClass("ticketList");
})

test('Matches Snapshot', () => {
    const tree = renderer.create(<TicketList ticketList={data}/>).toJSON();
    expect(tree).toMatchSnapshot();
})