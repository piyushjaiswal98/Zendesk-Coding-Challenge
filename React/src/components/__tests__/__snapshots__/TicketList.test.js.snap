// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches Snapshot 1`] = `
<div
  className="ticketList"
  data-testid="TicketList-1"
/>
`;
